### PR TITLE
[Th2 2422] Multiple dictionaries  access methods

### PR DESCRIPTION
**Pull request by Topru333**

### CODE REVIEW

File: src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java [hunk 1]
```diff
@@ -478,6 +501,113 @@ public static CommonFactory createFromKubernetes(String namespace, String boxNam
         }
     }
 
+    @Override
+    public InputStream loadDictionary() {
+        try {
+            Path dictionaryFolder = getPathToDictionaryAliasesDir();
+            LOGGER.debug("Loading dictionary from folder: {}", dictionaryFolder);
+            Set<Path> dictionaries = null;
+            if (Files.exists(dictionaryFolder) && Files.isDirectory(dictionaryFolder)) {
```
[the remark this answers is not in the document]
Comment (ivandruzhinin) [reply src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java:510]: Also probably we should check paths only at init

File: src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java [hunk 2]
```diff
@@ -478,6 +501,113 @@ public static CommonFactory createFromKubernetes(String namespace, String boxNam
...
+                try (Stream<Path> files = Files.list(dictionaryFolder)) {
+                    files.filter(Files::isRegularFile).forEach(dictionary -> {
+                        if (Files.isRegularFile(dictionary)) {
+                            dictionaries.add(FilenameUtils.removeExtension(dictionary.getFileName().toString()));
+                        }
+                    });
+                }
+            }
+            return dictionaries;
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java:545]:
```suggestion
            if (!Files.isDirectory(dictionaryFolder)) {
                return Set.of();
            }
            
            try (Stream<Path> files = Files.list(dictionaryFolder)) {
                return files.filter(Files::isRegularFile)
                        .map((dictionary) -> FilenameUtils.removeExtension(dictionary.getFileName().toString()))
                        .collect(Collectors.toSet());
            }
```

File: src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java [hunk 3]
```diff
@@ -478,6 +501,113 @@ public static CommonFactory createFromKubernetes(String namespace, String boxNam
...
+        }
+    }
+
+    @Override
+    public InputStream loadDictionary(String alias) {
+        var dictionaryName = alias.toLowerCase();
+        try {
+            Path dictionaryFolder = getPathToDictionaryAliasesDir();
+            LOGGER.debug("Loading dictionary alias ({}) from folder: {}", alias, dictionaryFolder);
+            if (Files.exists(dictionaryFolder) && Files.isDirectory(dictionaryFolder)) {
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java:557]:
```suggestion
            if (Files.isDirectory(dictionaryFolder)) {
```

File: src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java [hunk 4]
```diff
@@ -478,6 +501,113 @@ public static CommonFactory createFromKubernetes(String namespace, String boxNam
...
+                try (Stream<Path> stream = Files.list(dictionaryFolder)) {
+                    for (Path pathToAlias : stream.filter(Files::isRegularFile).collect(Collectors.toList())) {
+                        String fileNameWithoutExt = FilenameUtils.removeExtension(pathToAlias.getFileName().toString());
+                        if (fileNameWithoutExt.toLowerCase().equals(dictionaryName)) {
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java:561]:
```suggestion
                        if (fileNameWithoutExt.equalsIgnoreCase(alias)) {
```

File: src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java [hunk 5]
```diff
@@ -478,6 +501,113 @@ public static CommonFactory createFromKubernetes(String namespace, String boxNam
...
+    @Override
+    public InputStream readDictionary() {
+        return readDictionary(DictionaryType.MAIN);
+    }
+
+    @Override
+    public InputStream readDictionary(DictionaryType dictionaryType) {
+        try {
+            List<Path> dictionaries = null;
+            Path typeFolder = dictionaryType.resolvePathFrom(getPathToDictionaryTypesDir());
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java:583]: I think this method should also look for dictionary with alias = type for backward compatibility

File: src/test/kotlin/com/exactpro/th2/common/schema/TestDictionaryLoad.kt [hunk 6]
```diff
@@ -39,4 +42,40 @@ class TestDictionaryLoad {
...
+        Assertions.assertThrows(IllegalStateException::class.java) {
+            factory.loadDictionary()
+        }
+
+        val customSettings = FactorySettings().apply { dictionaryAliasesDir = Path.of("src/test/resources/test_load_dictionaries/single_dictionary") }
+        val customFactory = CommonFactory(customSettings)
+
+        customFactory.loadDictionary().use {
+            assert(String(it.readAllBytes()) == "test file")
+        }
```
Comment (ivandruzhinin) [inline src/test/kotlin/com/exactpro/th2/common/schema/TestDictionaryLoad.kt:78]: I don't think that we should mix checking of valid and invalid cases in a single test method

**Review by ivandruzhinin — CHANGES_REQUESTED** [review]

File: src/main/java/com/exactpro/th2/common/schema/dictionary/DictionaryType.java [hunk 7]
```diff
@@ -47,7 +47,7 @@
      * @param parent parent directory
      * @return Path to a directory which contains dictionaries with this type
      */
-    public Path getDictionary(Path parent) {
+    public Path resolvePathFrom(Path parent) {
```
[the remark this answers is not in the document]
Comment (Topru333) [reply src/main/java/com/exactpro/th2/common/schema/dictionary/DictionaryType.java:50]: Gonna change it back but i could not resist myself from change this
Dictionary.getDictionary(dictionary).resolve(dictionary)

File: src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java [hunk 8]
```diff
@@ -105,7 +116,8 @@
     private static final String GENERATED_CONFIG_DIR_NAME = "generated_configs";
 
     private final Path custom;
-    private final Path dictionariesDir;
+    private final Path dictionaryTypesDir;
+    private final Path dictionaryAliasesDir;
     private final Path oldDictionariesDir;
```
[the remark this answers is not in the document]
Comment (Topru333) [reply src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java:121]: Yes, we gonna remove it in common 4 but right now we need to support them

File: src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Topru333) [reply src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java:510]: There no guarantee that default folder exist on init, there cases without dictionary as a result no folder for them

File: src/main/java/com/exactpro/th2/common/schema/factory/AbstractCommonFactory.java [hunk 9]
```diff
@@ -524,48 +524,41 @@ public CradleManager getCradleManager() {
     }
 
     /**
+     * Read single dictionary
```
Comment (Topru333) [inline src/main/java/com/exactpro/th2/common/schema/factory/AbstractCommonFactory.java:527]: Better to form different defenition

File: src/main/kotlin/com/exactpro/th2/common/schema/factory/FactorySettings.kt [hunk 10]
```diff
@@ -45,7 +45,8 @@ data class FactorySettings @JvmOverloads constructor(
     var prometheus: Path? = null,
     var boxConfiguration: Path? = null,
     var custom: Path? = null,
-    var dictionariesDir: Path? = null,
+    @Deprecated("Will be removed in future releases") var dictionaryTypesDir: Path? = null,
+    var dictionaryAliasesDir: Path? = null,
     var oldDictionariesDir: Path? = null) {
```
Comment (Topru333) [inline src/main/kotlin/com/exactpro/th2/common/schema/factory/FactorySettings.kt:50]: Better to add Deprecatd annotation too

File: src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java [hunk 11]
```diff
@@ -478,6 +500,120 @@ public static CommonFactory createFromKubernetes(String namespace, String boxNam
...
+    @Override
+    public Set<String> getDictionaryAliases() {
+        Path dictionaryFolder = getPathToDictionaryAliasesDir();
+        try {
+            if (!Files.isDirectory(dictionaryFolder)) {
+                return Set.of();
+            }
+
+            try (Stream<Path> files = Files.list(dictionaryFolder)) {
+                return files.filter(Files::isRegularFile).map(dictionary -> FilenameUtils.removeExtension(dictionary.getFileName().toString())).collect(Collectors.toSet());
```
Comment (andreydrobynin) [inline src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java:538]: I think it's more readable:
```
return files
        .filter(Files::isRegularFile)
        .map(dictionary -> FilenameUtils.removeExtension(dictionary.getFileName().toString()))
        .collect(Collectors.toSet());
```
What do you think?
I think it's OK if we have short chain like `files.filter(Files::isRegularFile).collect(Collectors.toList())`.

File: src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java [hunk 12]
```diff
@@ -53,29 +53,36 @@
...
 
+import static com.exactpro.th2.common.schema.util.ArchiveUtils.getGzipBase64StringDecoder;
 import static java.util.Collections.emptyMap;
 import static java.util.Objects.requireNonNull;
 import static java.util.Objects.requireNonNullElse;
 import static org.apache.commons.lang3.ObjectUtils.defaultIfNull;
 
 /**
- * Default implementation for {@link AbstractCommonFactory}
+ * Default (platform specific) implementation for {@link AbstractCommonFactory}
```
Comment (andreydrobynin) [inline src/main/java/com/exactpro/th2/common/schema/factory/CommonFactory.java:85]: Why is it needed?

**Review by andreydrobynin — APPROVED** [review]

**Review by ivandruzhinin — DISMISSED** [review]

**Review by ivandruzhinin — APPROVED** [review]

**Pull request merged**